time_span/time_span.go

```diff
@@ -2,28 +2,210 @@ package time_span
+	// iff start1 <= end2 AND start2 <= end1.
+	return (t1.Start.Before(t2.End) || t1.Start.Equal(t2.End)) &&
+		(t2.Start.Before(t1.End) || t2.Start.Equal(t1.End))
+}
+
+// TODO: overlap between open and closed?
+
+// An open interval (start,end)
+type OpenInterval BaseInterval
+
+func NewOpenInterval(start time.Time, end time.Time) OpenInterval {
+	if start.After(end) {
+		start, end = end, start
+	}
+
+	return OpenInterval{
+		Start: start,
+		End:   end,
+	}
+}
+
+// Extend an interval by "delta" in each direction (thus it gets 2*delta longer.)
+func (t OpenInterval) Expand(delta time.Duration) OpenInterval {
+	return OpenInterval{
+		Start: t.Start.Add(time.Duration(-1) * delta),
+		End:   t.End.Add(delta),
+	}
+}
+
+// Return the portion of this interval that lies entirely within another open interval
+func (t OpenInterval) Intersect(t2 OpenInterval) OpenInterval {
+	return OpenInterval{
+		Start: MaxTime(t.Start, t2.Start), // later of the two starts
+		End:   MinTime(t.End, t2.End),     // earlier of the two ends
+	}
+}
+
+func (t OpenInterval) Duration() time.Duration {
+	if t.End.After(t.Start) {
+		return t.End.Sub(t.Start)
+	} else {
+		return time.Duration(0)
+	}
+}
+
+// Determines whether the interval includes the given query.
+func (t OpenInterval) Includes(query time.Time) bool {
+	return t.Start.Before(query) && query.Before(t.End)
+}
+
+// TODO: overlap between open and closed?
+
+// Determine whether this interval overlaps with another open interval
+// If one of them is empty, this should still return false.
+func (t1 OpenInterval) Overlaps(t2 OpenInterval) bool {
+	// Need C such that s1 < C < e1 and s2 < C < e2, so it is necessary
+	// that s1 < e2 and s2 < e1, but also that both of them are nonempty.
+	return t1.Start.Before(t2.End) && t2.Start.Before(t1.End) &&
+		t1.Start.Before(t1.End) && t2.Start.Before(t2.End)
+}
+
+// An half-open interval [start,end)
+type HalfOpenInterval BaseInterval
+
+func NewHalfOpenInterval(start time.Time, end time.Time) HalfOpenInterval {
+	return HalfOpenInterval{
+		Start: start,
+		End:   end,
+	}
+}
+
+func (span HalfOpenInterval) Duration() time.Duration {
+	if span.End.After(span.Start) {
+		return span.End.Sub(span.Start)
+	} else {
+		return time.Duration(0)
+	}
+}
+
+// Determines whether the interval includes the given query.
+func (span HalfOpenInterval) Includes(query time.Time) bool {
+	return (span.Start.Equal(query) || span.Start.Before(query)) && query.Before(span.End)
+}
+
+func (t1 HalfOpenInterval) Overlaps(t2 HalfOpenInterval) bool {
+	// There's some time C such that start1 <= C < end1 and start2 <= C < end2
+	// iff start1 < end2 AND start2 < end1.
+	return t1.Start.Before(t2.End) && t2.Start.Before(t1.End)
+}
+
+// Returns true if t2 is entirely contained within t1
+func (t1 HalfOpenInterval) Contains(t2 HalfOpenInterval) bool {
+	return (t1.Start.Before(t2.Start) || t1.Start.Equal(t2.Start)) &&
+		(t1.End.After(t2.End) || t1.End.Equal(t2.End))
+}
```

Comment:
Passing different representations of the empty interval to this implementation can yield different results:
```
nonEmpty := [1, 2)
empty1 := [2, 1)
empty2 := [3, 0)
empty3 := [4, 3)
nonEmpty.Contains(empty1) // returns true
nonEmpty.Contains(empty3) // returns false
empty1.Contains(empty2) // returns true
empty1.Contains(empty3) // returns false
```